misfit-preprocessor.py

```diff
@@ -0,0 +1,403 @@
+
+# Second subplot - Scaled responses
+ax2.set_title("Scaled responses")  # Fixed typo: "set_tiel" -> "set_title"
+ax2.set_xlabel("Realization")
+ax2.set_ylabel("Response")
+for i in range(Y_std.shape[0]):
+    ax2.plot(Y_std[i, :], alpha=0.3)
+
+# Adjust layout to prevent overlap
+plt.tight_layout()
+
+_, s_std, _ = np.linalg.svd(Y_std.T, full_matrices=False)
+
+var_ratio_std = np.cumsum(s_std**2) / np.sum(s_std**2)
+
+threshold = 0.95
+# TODO: There's a PR that improves the way we count pca components
+nr_components_std = max(len([1 for i in var_ratio_std[:-1] if i < threshold]), 1)
+
+fig, ax = plt.subplots()
+ax.set_title("Variance explained")
+ax.plot(range(1, len(var_ratio_std) + 1), var_ratio_std, linestyle="-", linewidth=2)
+ax.axvline(nr_components_std, linestyle="--", alpha=0.7, label="95% variance")
+ax.annotate(
+    f"{nr_components_std} components\nexplain 95% var", xy=(nr_components_std + 1, 0.7)
+)
+ax.legend()
+
+correlation = spearmanr(Y_std.T).statistic
+
+linkage_matrix = linkage(correlation, "average", "euclidean")
+
+cluster_labels = fcluster(linkage_matrix, nr_components_std, criterion="maxclust")
+
+fig, ax = plt.subplots()
+ax.hist(
+    cluster_labels, bins=range(1, nr_components_std + 2), alpha=0.7, edgecolor="black"
+)
+ax.set_xlabel("Cluster Label")
+ax.set_ylabel("Number of Observations")
+ax.set_title(f"Distribution of Cluster Assignments (k={nr_components_std})")
+ax.set_xticks(range(1, nr_components_std + 1))
+ax.grid(True, alpha=0.3)
+
+
+# %% [markdown]
+# # Alternative method using Kmeans and misfits
+
+
+# %%
+def find_optimal_clusters_elbow(data):
+    """
+    Find optimal number of clusters using the "elbow method"
+
+    The elbow method works like this:
+    1. Try different numbers of clusters (2, 3, 4, etc.)
+    2. For each number, measure how "tight" the clusters are (inertia)
+    3. Plot inertia vs number of clusters - it usually looks like an arm
+    4. The "elbow" (bend) in this curve is the optimal number
+
+    Think of it like: "How many groups do I need before adding more groups
+    doesn't help much?"
+    """
+    max_k = min(10, len(data) // 5)  # Don't try more clusters than we have data points
+    K_range = range(2, max_k + 1)  # Try 2, 3, 4, ... clusters
+
+    inertias = []  # This will store how "spread out" each clustering solution is
+    for k in K_range:
+        kmeans = KMeans(n_clusters=k, random_state=42, n_init=10)
+        kmeans.fit(data.reshape(-1, 1))
+        inertias.append(
+            kmeans.inertia_
+        )  # inertia = sum of distances from points to cluster centers
+
+    # Find the "elbow" - where the improvement starts slowing down
+    if len(inertias) >= 3:
+        diffs = np.diff(inertias)  # How much inertia decreases each step
+        diff2 = np.diff(diffs)  # How much the decrease rate changes
+        elbow_idx = np.argmax(diff2)  # Find biggest change in decrease rate
+        optimal_k = K_range[elbow_idx + 1]
```

Comment:
A very naive question. Don't we want to select such a k that sum of inertias is minimum out of all?